src/java/org/apache/cassandra/service/accord/txn/TxnKeyRead.java

```diff
@@ -0,0 +1,246 @@
+    {
+        return read.key();
+    }
+
+    @Override
+    TxnNamedRead[] newArray(int size)
+    {
+        return new TxnNamedRead[size];
+    }
+
+    @Override
+    public Keys keys()
+    {
+        return txnKeys;
+    }
+
+    @Override
+    public ConsistencyLevel cassandraConsistencyLevel()
+    {
+        return cassandraConsistencyLevel;
+    }
+
+    @Override
+    public Read slice(Ranges ranges)
+    {
+        return intersecting(itemKeys.slice(ranges));
+    }
+
+    @Override
+    public Read intersecting(Participants<?> participants)
+    {
+        return intersecting(itemKeys.intersecting(participants));
+    }
+
+    private Read intersecting(Keys select)
+    {
+        Keys keys = itemKeys.intersecting(select);
+        List<TxnNamedRead> reads = new ArrayList<>(keys.size());
+
+        for (TxnNamedRead read : items)
+            if (keys.contains(read.key()))
+                reads.add(read);
+
+        return createTxnRead(reads, txnKeys.intersecting(select), cassandraConsistencyLevel);
+    }
+
+    @Override
+    public Read merge(Read read)
+    {
+        List<TxnNamedRead> reads = new ArrayList<>(items.length);
+        Collections.addAll(reads, items);
+
+        for (TxnNamedRead namedRead : (TxnKeyRead) read)
+            if (!reads.contains(namedRead))
+                reads.add(namedRead);
+
+        return createTxnRead(reads, txnKeys.with((Keys)read.keys()), cassandraConsistencyLevel);
+    }
+
+    @Override
+    public AsyncChain<Data> read(Seekable key, SafeCommandStore safeStore, Timestamp executeAt, DataStore store)
+    {
+        List<AsyncChain<Data>> results = new ArrayList<>();
+        forEachWithKey((PartitionKey) key, read -> results.add(read.read(cassandraConsistencyLevel, executeAt)));
+
+        if (results.isEmpty())
+            // Result type must match everywhere
+            return AsyncChains.success(new TxnData());
+
+        if (results.size() == 1)
+            return results.get(0);
+
+        return AsyncChains.reduce(results, Data::merge);
+    }
+
+    @Override
+    public Kind kind()
+    {
+        return Kind.key;
+    }
+
+    @Simulate(with = MONITORS)
+    public static final TxnReadSerializer<TxnKeyRead> serializer = new TxnReadSerializer<TxnKeyRead>()
+    {
+        @Override
+        public void serialize(TxnKeyRead read, DataOutputPlus out, int version) throws IOException
+        {
+            KeySerializers.keys.serialize(read.txnKeys, out, version);
+            serializeArray(read.items, out, version, TxnNamedRead.serializer);
+            serializeNullable(read.cassandraConsistencyLevel, out, version, consistencyLevelSerializer);
+        }
+
+        @Override
+        public TxnKeyRead deserialize(DataInputPlus in, int version) throws IOException
+        {
+            Keys keys = KeySerializers.keys.deserialize(in, version);
+            TxnNamedRead[] items = deserializeArray(in, version, TxnNamedRead.serializer, TxnNamedRead[]::new);
```

Comment:
can we pass the Key through to the deserialiser, so it doesn't need to find the `TableMetadata` again?

Reply:
If we're looking at these serialisers and classes, it would be nice also to remove the String[] from `TxnDataName` and use a simple integer before we ossify the serialization format, if you don't mind including it in scope?

Reply:
I have a follow up patch that will do a bit more and fix some issues with how the ranges are passed to `TxnRangeRead` as part of making it live migratable. Can we do the switch from string to in there?

When you mean pass keys through so it doesn't need to find `TableMetadata` what do you mean? Keys I think only contains `PartitionKey` with a `TableId` not a `TableMetadata`.

Reply:
Where do you see that? I looked at the `PartitionKey` deserializer and it does `TableId.deserialize(in).intern();`, `TableId.deserialize` and the constructor don't do anything special.

Intern does
```
    public TableId intern()
    {
        TableId interned = internCache.putIfAbsent(this, this);
        return interned == null ? this : interned;
    }
```

I think it would actually easier to construct the `Keys` after deserializing the reads which should contain all the Keys that are in `Keys` in the first place. That is how we construct the `Keys` in the first place.

Reply:
Oh this is so weird. There are two sets of keys here. One that is `AbstractKeySorted` and another that is in `TxnKeyRead`. One of them is all the keys for the txn, the other is all the keys extracted from the actual reads. I don't get why we would want the read to include all the keys for the entire transaction? 

Going to try eliminating that and see what happens...